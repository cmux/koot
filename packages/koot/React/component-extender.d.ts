/* eslint-disable @typescript-eslint/no-explicit-any */
// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference path="../global.d.ts" />

import {
    // ComponentType,
    ReactNode,
    FC,
    ComponentClass,
    Component,
    ComponentState,
    RefObject,
} from 'react';
import { Store, Dispatch } from 'redux';
import {
    MapStateToPropsParam,
    MapDispatchToPropsNonObject,
    MapDispatchToPropsParam,
    MergeProps,
    Options as ReactReduxOptions,
} from 'react-redux';

// ============================================================================

declare module 'kootExtendHOC';

// ============================================================================

interface ExtendComponent<P, S> {
    (wrappedComponent: FC<ExtendedProps & P>): ComponentClass<
        ExtendedProps & P
    >;
    (wrappedComponent: ComponentClass<any, any>): HOC<ExtendedProps & P, S>;
    // (wrappedComponent: FC<ExtendedProps & P>): ComponentClass<
    //     ExtendedProps & P
    // >;

    // (
    //     wrappedComponent:
    //         | FC<ExtendedProps & P>
    //         | ComponentClass
    //         | ComponentClass<ExtendedProps & P>
    // ): ComponentClass<ExtendedProps & P>;

    // (wrappedComponent: ComponentType<any>): HOC<ExtendedProps & P, S>;
    // (wrappedComponent: ComponentClass<any, any>): ComponentClass<
    //     JSX.LibraryManagedAttributes<wrappedComponent, P> & ExtendedProps & P,
    //     S
    // >;
}
class HOC extends Component {}

export function extend<P = {}, S = ComponentState>(options: {
    // connect?: Connect;
    connect?:
        | true
        | MapStateToPropsParam<any, P, S>
        | [null | undefined | MapStateToPropsParam<any, P, S>]
        | [
              null | undefined | MapStateToPropsParam<any, P, S>,
              (
                  | null
                  | undefined
                  | MapDispatchToPropsNonObject<any, P>
                  | MapDispatchToPropsParam<any, P>
              )
          ]
        | [
              null | undefined | MapStateToPropsParam<any, P, S>,
              (
                  | null
                  | undefined
                  | MapDispatchToPropsNonObject<any, P>
                  | MapDispatchToPropsParam<any, P>
              ),
              null | undefined | MergeProps<any, any, P, any>
          ]
        | [
              null | undefined | MapStateToPropsParam<any, P, S>,
              (
                  | null
                  | undefined
                  | MapDispatchToPropsNonObject<any, P>
                  | MapDispatchToPropsParam<any, P>
              ),
              null | undefined | MergeProps<any, any, P, any>,
              null | undefined | ReactReduxOptions<S, any, P>
          ];
    /** 提供页面的 title 和 meta 标签信息 */
    pageinfo?: Pageinfo | ExtendPageinfoFunction;
    data?: DataFetchFunction | ExtendData;
    styles?: KootComponentStyleObject;
    /**
     * 控制 SSR 行为
     * @default true
     */
    ssr?: ReactNode | boolean;
    /** 组件名 (仅供调试用) */
    name?: string;
}): ExtendComponent<P, S>;

export type Pageinfo = {
    /** 要设置的页面标题 */
    title?: string;
    /** 要设置的 `<meta>` 标签列表，需要元素为 `{[name]: value}` 对象的数组 */
    metas?: Array<MetaObject>;
};
export type MetaObject = {
    [metaKey: string]: string | undefined;
};

type ExtendPageinfoFunction = (state: any, props: any) => Pageinfo;

type DataFetchFunction = (
    store: Store,
    renderProps: any,
    dispatch: Dispatch
) => Promise<any>;

type DataCheckFunction = (state: any, renderProps: any) => boolean;

type ExtendData = {
    fetch?: DataFetchFunction;
    check?: DataCheckFunction;
};

//

/** extend 高阶组件向目标组件注入的 props */
export interface ExtendedProps {
    /** 包含父组件传入的 `className` 和 `extend()` 高阶组件注入的 `className`（如果在 `extend()` 高阶组件中传入了 `styles` 属性）*/
    className?: string;
    /**
     * `extend()` 高阶组件注入的 `className`
     * - 不包含父组件传入的 `className`
     * - 仅在 `extend()` 高阶组件中传入了 `styles` 属性时才会存在
     */
    readonly 'data-class-name'?: string;
    /**
     * 可用来手动触发页面信息更新，语法与 `connect()` 高阶组件的 `pageinfo` 的函数用法相同
     * - 如果没有传入参数，会使用 `extend()` 高阶组件中的 `pageinfo` 的值
     * - 仅在 `extend()` 高阶组件中使用 `pageinfo` 属性时才会存在
     */
    readonly updatePageinfo?: ExtendedPropsUpdatePageinfo;
    readonly forwardedRef?: RefObject<any>;
}
type ExtendedPropsUpdatePageinfo = (
    newPageinfo?: Pageinfo | ExtendPageinfoFunction
) => void;

interface RenderProps extends ExtendedProps {
    readonly location?: Record<string, any>;
    readonly params?: Record<string, any>;
    readonly route?: Record<string, any>;
    readonly routeParams?: Record<string, any>;
    readonly router?: Record<string, any>;
    readonly routes?: Array<Record<string, any>>;
}
