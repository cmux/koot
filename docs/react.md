# React

### 高阶组件 `extend`

_Koot.js_ 为 _React_ 提供高阶组件 _extend()_，可赋予目标组件**CSS 命名空间**、**同构数据**、**更新页面信息**等能力。

针对 _同构/SSR_ 项目，同构渲染 CSS `<style>` 标签、异步数据请求、页面标题与 `<meta>` 标签等信息，均通过该高阶组件实现。

#### 使用

```javascript
import { extend } from 'koot';

class HomePage extends React.Component {
    // ...
}

export default extend({
    // ...配置对象
})(HomePage);
```

也可以作为装饰器使用

```javascript
import { extend } from 'koot';

@extend({
    // ...配置对象
})
class HomePage extends React.Component {
    // ...
}

export default HomePage;
```

针对函数组件 (Funtional Component) 的使用

```javascript
import { extend } from 'koot';

const HomePage = extend({
    // ...配置对象
})(() => {
    // ...
});

export default HomePage;
```

**TypeScript**

```typescript

// 函数方式定义
extend<ComponentProps>(extendOptions)(FunctionalComponent)

// 装饰器方式定义
@extend(extendOptions)
ComponentClass
```

在 TypeScript 中的写法请参见 [TypeScript 开发/TSX 代码示例](/typescript?id=tsx-代码示例)

##### 参数对象属性

_Function_ `connect`

-   传入 `react-redux` 的高阶组件 `connect` 的回调函数。
-   传入多个参数的方法请见下文示例。

_Object_ `styles` 组件 CSS

-   一般情况下，该属性为引用 `css` `less` 或 `sass` 文件。
-   为该组件提供或更新 `props.className`
    -   如果父级组件传入了 `props.className`，该子组件的 `props.className` 会包含父级传入的值以及引用的组件 CSS 的样式名
    -   如果父级组件没有传入 `props.className`，该子组件的 `props.className` 仅为引用的组件 CSS 的样式名
-   该组件渲染时，这些 CSS 会作为 `<style>` 标签添加到 `<head>` 中。
-   该对象需要提供以下属性：_string_ `wrapper` 和 _string_ `css`
-   有关 CSS 的相关开发指南请查阅 [CSS](/css)

`pageinfo` 修改页面 title 和 meta 标签

-   该组件渲染时，这些信息会更新到 `<head>` 中。
-   该属性有两种用法：
    -   _Object_ `pageinfo`
        -   _string_ `pageinfo.title`
        -   _Array<MetaObject>_ `pageinfo.metas`
    -   _Function_ `pageinfo`
        <br>`pageinfo(state, renderProps)`
        -   参数
            -   _Object_ `state` 当前 Redux state
            -   _Object_ `renderProps` 同构对象
        -   返回值：_Object_
            -   _string_ `title` 新的页面标题
            -   _Array_ `metas` 新的页面 meta 标签信息
-   如果没有提供 `title`，会默认采用项目配置中的 `name` 值，如果该值也未提供，则会默认使用 `package.json` 的 `name` 属性

`data` 同构数据相关

-   该属性有两种用法：
    -   _Object_ `data`
        -   _Function_ `data.fetch` 获取数据的方法函数
            <br>`fetch(state, renderProps, dispatch)`
            -   参数
                -   _Object_ `state` 当前 Redux state
                -   _Object_ `renderProps` 同构对象
                -   _Function_ `dispatch` Redux dispatch 方法
            -   返回值：_Promise_
        -   _Function_ `data.check` 检查数据是否已准备就绪
            <br>`check(state, props)`
            -   参数
                -   _Object_ `state` 当前 Redux state
                -   _Object_ `renderProps` 同构对象
            -   返回值：_Boolean_
    -   _Function_ `data`
        <br>为 _Function_ 时，同 `data.fetch`
        <br>使用该方法时，需要自行编写检查数据的代码，推荐写在 redux action 中

`ssr` 仅作用于 SSR 项目的服务器端：控制该组件的 SSR 行为

-   默认值：`true`
-   _Boolean_ `ssr`
    <br>该组件是否需要 SSR。
    <br>`false` 时，SSR 阶段不会渲染该组件，最终在 HTML 结果中不会出现相应的 HTML 代码。
-   _string_|_ReactComponent_ `ssr`
    <br>在 SSR 时渲染指定的内容或组件

##### 示例：使用所有参数

```javascript
import { extend } from 'koot';
import { fetchUser } from '@api/user';

import styles from './styles.less';

@extend({
    // connect 仅传入第一个参数
    connect: (state) => ({
        user: state.user || {},
    }),
    // connect 传入多个参数
    connect: [
        (state) => ({
            user: state.user || {},
        }),
        (dispatch, ownProps) => ({
            getUserData: () =>
                dispatch({ type: 'GET_USER_DATA', id: ownProps.id }),
        }),
    ],

    //
    styles,

    // pageinfo 使用 Object 方式
    pageinfo: {
        title: __('pages.home.title'),
        metas: [{ name: 'description', content: __('pages.home.description') }],
    },
    // pageinfo 使用 Function 方式
    pageinfo: (state, renderProps) => ({
        title: __('pages.home.title'),
        metas: [{ name: 'description', content: __('pages.home.description') }],
    }),

    // data 使用 Object 方式
    data: {
        fetch: (state, renderProps, dispatch) => {
            return dispatch(fetchUser());
        },
        check: (state, renderProps) => {
            return (
                typeof renderProps.user === 'object' &&
                typeof renderProps.user.id !== 'undefined'
            );
        },
    },
    // data 使用 Function 方式
    data: (state, renderProps, dispatch) => {
        return dispatch(fetchUser());
    },

    // 不进行 SSR
    ssr: false,
    // SSR 时渲染该组件
    ssr: <div>Loading...</div>,
})
class HomePage extends React.Component {
    // ...
}

export default HomePage;
```

##### 组件获得新的 props

使用 _extend()_ 高阶组件封装后，目标组件将会获得以下新的 `props`

| 属性名            | 类型              | 解释                                                                                                                                                                                                                                                                                                                   |
| ----------------- | ----------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `className`       | `string`          | 如果通过 `extend()` 高阶组件封装并指定了 `styles` 属性：<br>父组件传入的 `className` 以及经过 hash 后本组件的 `className`<br><br>否则：<br>父组件传入的 `className` 属性                                                                                                                                               |
| `data-class-name` | `string`          | 经过 hash 后本组件的 className<br>与 `className` 不同，该属性仅为本组件的组件 CSS 的 `className`                                                                                                                                                                                                                       |
| `updatePageinfo`  | `Function`        | 可用来手动触发页面信息更新<br><br>可传入 _Object_ 或 _Function_，语法与 `connect()` 高阶组件的 `pageinfo` 的用法相同<br>如果没有传入参数，会使用 `extend()` 高阶组件中的 `pageinfo` 的值<br><br>⚠️ 仅在客户端环境下使用才有意义，服务器环境下该属性无效<br>⚠️ 仅在 `extend()` 高阶组件中使用 `pageinfo` 属性时才会存在 |
| `forwardedRef`    | `React.RefObject` | 使用该组件时如果传入了 `ref` 属性，该组件会出现 `forwardedRef` 属性用以转发/接收 `ref` 值                                                                                                                                                                                                                              |

---

### 组件热更新

在**开发环境**中，所有 React 组件均会默认启用热更新机制

> 所谓热更新，即在编辑器中修改并保存了代码文件后，无需刷新网页即可看到最新的结果。

⚠️ Koot.js 默认的热更新能力有以下限制：

-   仅影响扩展名为 `.jsx` `.tsx` 的文件
    -   非 React 组件文件请勿以 `.jsx` `.tsx` 作为文件扩展名
-   仅影响文件中输出 (`export`) 的组件，没有输出的组件不会有热更新

**有关 `memo()`**

出于某些限制，目前 _Koot.js_ 仍在使用 `react-hot-loader`，其作者已经表示不再维护，目前已知对 `memo()` 组件的热更新支持不佳。

为了缓解这一问题，在开发环境中，所有的 `memo()` 函数都会被**自动地**替换为组件本身，从而使热更新可用，但这样一来和生产环境地结果就有了一定的偏差，在开发时请注意。

⚠️ 在未来会更新为全新的 _Fast Refresh_ 机制，从而杜绝这一现象

### 同构对象

Koot.js 自定义的对象，其中包含以下属性/元素：

_TODO:_
