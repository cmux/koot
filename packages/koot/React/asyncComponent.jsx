/* eslint-disable react/no-unsafe */
// https://gist.github.com/acdlite/a68433004f9d6b4cbc83b5cc3990c194

const { Component } = require('react');

module.exports = (getComponent) =>
    class AsyncComponent extends Component {
        static Component = null;
        state = { Component: AsyncComponent.Component };

        componentWillMount() {
            if (!this.state.Component) {
                getComponent().then((Component) => {
                    AsyncComponent.Component = Component;
                    this.setState({ Component });
                });
            }
        }
        render() {
            const { Component } = this.state;
            if (Component) {
                return <Component {...this.props} />;
            }
            return null;
        }
    };
