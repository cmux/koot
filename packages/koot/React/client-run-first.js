import 'regenerator-runtime/runtime';
import { syncHistoryWithStore } from 'react-router-redux';
import * as portionConfig from '__KOOT_PROJECT_CONFIG_PORTION_CLIENT_PATHNAME__';
import { STORE, HISTORY, DEV_NATIVE_CONSOLE } from '../defaults/defines-window';
import validateReduxConfig from './validate/redux-config';
import History from './history';

window[STORE] = ((reduxConfig = {}) => {
    if (typeof reduxConfig.factoryStore === 'function')
        return reduxConfig.factoryStore();
    if (typeof reduxConfig.store === 'object') return reduxConfig.store;
    return {};
})(validateReduxConfig(portionConfig.redux));

window[HISTORY] = syncHistoryWithStore(History, window[STORE]);

// console filter
// https://stackoverflow.com/questions/6659312/is-there-a-way-to-filter-output-in-google-chromes-console
if (__DEV__) {
    (() => {
        // eslint-disable-next-line no-console
        console.info('ðŸ’• [DEV] Thanks for using Koot.js');

        window.console = window.console || { log: 0 };

        const nativeConsole = {
            log: window.console.log,
            info: window.console.info,
            error: window.console.error,
        };
        window[DEV_NATIVE_CONSOLE] = nativeConsole;

        const lastLog = {
            WDS: {
                time: undefined,
                log: undefined,
            },
            HMR: {
                time: undefined,
                log: undefined,
            },
        };
        const throttleHotLog = (level) => {
            window.console[level] = function () {
                try {
                    const args = Array.from(arguments);

                    // throttle `WDS` and `HMR` logs
                    const needThrottle = (type) => {
                        const regex = new RegExp(`^\\[${type}\\]`);
                        if (!args.some((arg) => regex.test(arg))) return false;

                        const { [type]: last } = lastLog;
                        const now = Date.now();
                        const nowLogStr = args.join('|||');
                        const { time: lastTime, log: lastLogStr } = last;
                        lastLog[type] = {
                            time: now,
                            log: nowLogStr,
                        };
                        // console.warn(type, last, now - lastTime)

                        if (
                            lastTime &&
                            now - lastTime < 50 &&
                            lastLogStr &&
                            lastLogStr === nowLogStr
                        )
                            return true;
                        return false;
                    };
                    if (needThrottle('WDS')) return;
                    if (needThrottle('HMR')) return;

                    nativeConsole[level].apply(window.console, arguments);
                } catch (e) {}
            };
        };
        // throttleHotLog('log')
        throttleHotLog('info');

        const warningShowed = {
            reactRouterV3: false,
            reactReduxV5: false,
        };
        window.console.error = function () {
            try {
                const args = Array.from(arguments);

                // filter out SSR not matching error
                // will fix this by using koot's own async component
                // SOONâ„¢
                if (
                    args.some((arg) =>
                        /^Warning: Did not expect server HTML to contain a <%s> in <%s>\.$/.test(
                            arg
                        )
                    )
                )
                    return;

                // filter out react-router v3` warnings/errors
                if (
                    (/^Warning: Legacy context API has been detected within a strict-mode tree/.test(
                        args[0]
                    ) &&
                        args.some((arg) => arg === 'Link')) ||
                    (/^Warning: Using UNSAFE_[a-zA-Z]+ in strict mode/.test(
                        args[0]
                    ) &&
                        /(Link)(\r|\n|,|\(|$)/.test(args[1]))
                ) {
                    if (!warningShowed.reactRouterV3) {
                        // eslint-disable-next-line no-console
                        console.info(
                            '\n[koot] Koot.js is now using `react-router` v3 which will be upgraded to newer version in future.\n\n'
                        );
                        warningShowed.reactRouterV3 = true;
                    }
                    return;
                }

                nativeConsole.error.apply(window.console, arguments);
            } catch (e) {}
        };
    })();
}
