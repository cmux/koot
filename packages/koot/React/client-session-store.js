import get from 'lodash/get';
import set from 'lodash/set';

import { getStore } from '../';
import { sessionStoreKey } from '../defaults/defines';
import filterState from '../libs/filter-state';

const configSessionStore = JSON.parse(process.env.KOOT_SESSION_STORE);

/** @type {Boolean} 当前是否可以/允许使用 sessionStore */
const able = (() => {
    if (__SERVER__) return false;
    if (!window.sessionStorage) return false;
    if (!configSessionStore) return false;

    if (configSessionStore === true) return true;
    if (configSessionStore === 'all') return true;

    return Boolean(
        typeof configSessionStore === 'object' &&
            !Array.isArray(configSessionStore)
    );
})();

/**
 * 保存当前 state 到 sessionStorage
 * @void
 */
export const save = () => {
    if (!able) return;

    /** @type {Object} 排除掉黑名单内的项目后的 state 对象 */
    const state = filterState(getStore().getState());

    let saveState = {};

    if (configSessionStore === true || configSessionStore === 'all') {
        saveState = state;
    } else if (
        typeof configSessionStore === 'object' &&
        !Array.isArray(configSessionStore)
    ) {
        // 根据配置对象存储
        const parse = (obj, accumulatedKey = '') => {
            Object.keys(obj).forEach((key) => {
                const currentAccumulatedKey =
                    accumulatedKey + `[${JSON.stringify(key)}]`;
                if (typeof obj[key] === 'object') {
                    parse(obj[key], currentAccumulatedKey);
                } else if (obj[key] === true) {
                    const value = get(state, currentAccumulatedKey);
                    if (typeof value !== 'undefined')
                        set(saveState, currentAccumulatedKey, value);
                }
            });
        };
        parse(configSessionStore);
        // console.log(configSessionStore, state, saveState);
    }

    sessionStorage.setItem(sessionStoreKey, JSON.stringify(saveState));

    return;
};

/**
 * 向 window.onunload 添加事件：保存 state
 * @void
 */
export const addEventHandlerOnPageUnload = () => {
    if (!able) return;
    window.addEventListener('unload', save);
    return;
};

/**
 * 从 sessionStorage 中读取 state
 * @returns {Object} 存储的 state
 */
export const load = () => {
    if (!able) return {};
    return JSON.parse(sessionStorage.getItem(sessionStoreKey) || '{}');
};
