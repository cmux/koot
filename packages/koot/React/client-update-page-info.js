/**
 * @typedef Pageinfo
 * @type {Object}
 * @property {string} [title] 标题
 * @property {Array<Object.<string, string>>} [metas] meta 标签信息，需要形式为 `{[name]: value}` 对象的数组
 */

// import isEqual from 'lodash/isEqual';

/**
 * 当前已注入的 meta 标签
 * @type {Array<HTMLElement>}
 */
let injectedMetaTags;

// meta 标签区域结尾的 HTML 注释代码
let nodeCommentEnd;

let inited = false;

/** @type {Pageinfo} */
const infoToChange = {
    title: undefined,
    metas: [],
};
let changeTimeout = undefined;

/**
 * _仅限客户端_
 *
 * 更新页面信息
 * @module
 * @param {string} title 新的标题
 * @param {Array<Object.<string, string>>} metas meta 标签信息，需要形式为 `{[name]: value}` 对象的数组
 */
export default (title, metas = []) => {
    if (__SERVER__) return;
    if (!__SPA__ && !inited) {
        setTimeout(() => {
            inited = true;
        });
        return;
    }

    // 判断 & 追加即将修改的内容
    /*
    // 如果新 meta 不存在于已有的 meta 列表，添加
    metas.forEach(metaNew => {
        if (!infoToChange.metas.length) infoToChange.metas.push(metaNew);
        if (
            !infoToChange.metas.every(metaExist => {
                console.log(metaNew, metaExist, isEqual(metaNew, metaExist));
                return isEqual(metaNew, metaExist);
            })
        ) {
            console.log('PUSH');
            infoToChange.metas.push(metaNew);
        }
    });
    */
    // 如果当前没有信息，设为当前信息
    if (
        !infoToChange.title &&
        (!Array.isArray(infoToChange.metas) || !infoToChange.metas.length)
    ) {
        infoToChange.title = title;
        infoToChange.metas = metas;
    }

    if (changeTimeout) return;

    changeTimeout = setTimeout(() => {
        doUpdate();
        infoToChange.title = undefined;
        infoToChange.metas = [];
        changeTimeout = undefined;
    });
};

//

const doUpdate = () => {
    const { title, metas } = infoToChange;

    // 替换页面标题
    if (typeof title !== 'undefined') document.title = title;

    // 替换 metas
    const head = document.getElementsByTagName('head')[0];
    getInjectedMetaTags().forEach((el) => head.removeChild(el));

    injectedMetaTags.forEach((el) => {
        if (el && el.parentNode) el.parentNode.removeChild(el);
    });
    injectedMetaTags = metas
        .filter((meta) => typeof meta === 'object')
        .map((meta) => {
            const el = document.createElement('meta');
            for (var key in meta) {
                el.setAttribute(key, meta[key]);
            }
            // el.setAttribute(__KOOT_INJECT_ATTRIBUTE_NAME__, '')
            if (nodeCommentEnd) {
                head.insertBefore(el, nodeCommentEnd);
            } else {
                head.appendChild(el);
            }
            return el;
        });
};

//

/**
 * 获取当前已注入的 meta 标签
 * @returns {Array<HTMLElement>}
 */
export const getInjectedMetaTags = () => {
    if (!Array.isArray(injectedMetaTags)) {
        const head = document.getElementsByTagName('head')[0];

        injectedMetaTags = [];
        // 移除所有在 KOOT_METAS 里的 meta 标签
        // 采用 DOM 操作的初衷：如果使用 innerHTML 的字符串替换方法，浏览器可能会全局重新渲染一次，造成“闪屏”
        const childNodes = head.childNodes;
        let meetStart = false;
        let meetEnd = false;
        let i = 0;
        while (!meetEnd && childNodes[i] instanceof Node) {
            const node = childNodes[i];
            if (node.nodeType === Node.COMMENT_NODE) {
                if (node.nodeValue === __KOOT_INJECT_METAS_START__)
                    meetStart = true;
                if (node.nodeValue === __KOOT_INJECT_METAS_END__) {
                    meetEnd = true;
                    nodeCommentEnd = node;
                }
            } else if (
                meetStart &&
                node.nodeType === Node.ELEMENT_NODE &&
                node.tagName === 'META'
            ) {
                injectedMetaTags.push(node);
            }
            i++;
        }
    }

    return injectedMetaTags;
};

/** _仅针对客户端_ 标记已初始化 */
export const markInited = () => {
    inited = true;
};
