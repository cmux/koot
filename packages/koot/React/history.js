// import history from "__KOOT_CLIENT_REQUIRE_HISTORY__"
import createHistory from '__KOOT_CLIENT_REQUIRE_CREATE_HISTORY__';
import { parsePath } from 'history/lib/PathUtils';
import isI18nEnabled from '../i18n/is-enabled';
import { REDUXSTATE } from '../defaults/defines-window';

/**
 * History Enhancer: use basename
 *
 * Original useBasename enhancer from history also override all read methods
 * `getCurrentLocation` `listenBefore` `listen`
 * But as Diablohu tested, when read methods overrided, if the route matched used async method to get component, would fail
 * that rendering blank page and no route match event fired
 * So we only overrid write methods here. And modify the first level path in routes object to `:localeId`
 *
 * @param {Function} createHistory
 * @returns {Object} History
 */
const kootUseBasename =
    (createHistory) =>
    (options = {}) => {
        const history = createHistory(options);
        const { basename } = options;

        const addBasename = (location) => {
            if (!location) return location;

            if (basename && location.basename == null) {
                if (
                    location.pathname
                        .toLowerCase()
                        .indexOf(basename.toLowerCase()) === 0
                ) {
                    location.pathname = location.pathname.substring(
                        basename.length
                    );
                    location.basename = basename;

                    if (location.pathname === '') location.pathname = '/';
                } else {
                    location.basename = '';
                }
            }

            return location;
        };

        const prependBasename = (location) => {
            if (!basename) return location;

            const object =
                typeof location === 'string' ? parsePath(location) : location;
            const pname = object.pathname;
            const normalizedBasename =
                basename.slice(-1) === '/' ? basename : `${basename}/`;
            const normalizedPathname =
                pname.charAt(0) === '/' ? pname.slice(1) : pname;
            const pathname = normalizedBasename + normalizedPathname;

            return {
                ...object,
                pathname,
            };
        };

        // Override all write methods with basename-aware versions.
        const push = (location) => history.push(prependBasename(location));

        const replace = (location) =>
            history.replace(prependBasename(location));

        const createPath = (location) =>
            history.createPath(prependBasename(location));

        const createHref = (location) =>
            history.createHref(prependBasename(location));

        const createLocation = (location, ...args) =>
            addBasename(
                history.createLocation(prependBasename(location), ...args)
            );

        return {
            ...history,
            push,
            replace,
            createPath,
            createHref,
            createLocation,
        };
    };

let historyClient;

const history = (() => {
    if (__CLIENT__) {
        if (!historyClient) {
            const initialState = window[REDUXSTATE] || {};
            const historyConfig = { basename: '/' };
            if (
                isI18nEnabled() &&
                process.env.KOOT_I18N_URL_USE === 'router' &&
                initialState.localeId
            ) {
                historyConfig.basename = `/${initialState.localeId}`;
                historyClient = kootUseBasename(createHistory)(historyConfig);
            } else if (!!process.env.KOOT_HISTORY_BASENAME) {
                const basename = process.env.KOOT_HISTORY_BASENAME.replace(
                    /^\//,
                    ''
                ).replace(/\/$/, '');
                historyConfig.basename = `/${basename}`;
                historyClient = kootUseBasename(createHistory)(historyConfig);
            } else {
                // historyClient = require("__KOOT_CLIENT_REQUIRE_HISTORY__")
                historyClient = createHistory();
            }
            // console.log({
            //     historyConfig,
            //     historyClient
            // })
        }
        return historyClient;
    }

    if (__SERVER__) {
        return undefined;
    }
})();

export default history;
