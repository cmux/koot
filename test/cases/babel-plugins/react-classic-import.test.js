const babel = require('@babel/core');
const pluginReactClassicImport = require('../../../packages/koot-webpack/loaders/babel/plugins/react-classic-import');

const codeBlock = `
const ttt = 'aaa';
console.log(ttt);
`;
const examples = {
    noImport: `
${codeBlock}
`,
    noImport2: `
let tt;
${codeBlock}
`,
    hasImport: `
import React from 'react';
${codeBlock}
`,
    hasImport2: `
import React, { memo } from 'react';
${codeBlock}
`,
    hasImport3: `
import { memo } from 'react';
${codeBlock}
`,
    hasRequire: `
const React = require('react');
${codeBlock}
`,
    hasRequire2: `
const { memo } = require('react');
${codeBlock}
`,
};

describe('Babel Plugin: react-classic-import', () => {
    for (const [name, codeOriginal] of Object.entries(examples)) {
        it(name, () => {
            const { code } = babel.transform(codeOriginal, {
                plugins: [pluginReactClassicImport],
            });
            expect(code).toMatchSnapshot();
        });
    }
});
