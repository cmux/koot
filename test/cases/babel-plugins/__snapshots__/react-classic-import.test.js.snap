// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Babel Plugin: react-classic-import hasImport 1`] = `
"import React from 'react';
const ttt = 'aaa';
console.log(ttt);"
`;

exports[`Babel Plugin: react-classic-import hasImport2 1`] = `
"import React, { memo } from 'react';
const ttt = 'aaa';
console.log(ttt);"
`;

exports[`Babel Plugin: react-classic-import hasImport3 1`] = `
"import React from \\"react\\";
import { memo } from 'react';
const ttt = 'aaa';
console.log(ttt);"
`;

exports[`Babel Plugin: react-classic-import hasRequire 1`] = `
"const React = require('react');

const ttt = 'aaa';
console.log(ttt);"
`;

exports[`Babel Plugin: react-classic-import hasRequire2 1`] = `
"import React from \\"react\\";

const {
  memo
} = require('react');

const ttt = 'aaa';
console.log(ttt);"
`;

exports[`Babel Plugin: react-classic-import noImport 1`] = `
"import React from \\"react\\";
const ttt = 'aaa';
console.log(ttt);"
`;

exports[`Babel Plugin: react-classic-import noImport2 1`] = `
"import React from \\"react\\";
let tt;
const ttt = 'aaa';
console.log(ttt);"
`;
